File: packages/picoipc/cpp/benchmarks/bench_throughput.cpp
#include "smh_q/ring.hpp"
#include <unistd.h>

#include <algorithm>
#include <atomic>
#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <string>
#include <thread>
#include <vector>

using Clock = std::chrono::high_resolution_clock;

static std::uint64_t bench_now_ns() {
  return static_cast<std::uint64_t>(
      std::chrono::duration_cast<std::chrono::nanoseconds>(Clock::now().time_since_epoch())
          .count());
}

static std::uint32_t bench_slot_bytes(int payload_bytes) {
  return static_cast<std::uint32_t>((payload_bytes + 8) < 64 ? 64 : (payload_bytes + 8));
}

int main(int argc, char** argv) {
  const int count = (argc > 1) ? std::atoi(argv[1]) : 100'000;
  const int payload_bytes = (argc > 2) ? std::atoi(argv[2]) : 64;

  const std::string name = "/smh_q_bench_tp_" + std::to_string(::getpid());
  smh_q::Ring::unlink(name);

  smh_q::Ring::Config cfg;
  cfg.name = name;
  cfg.slot_count = 256;
  cfg.slot_size = bench_slot_bytes(payload_bytes);

  smh_q::Ring producer(cfg, true);
  smh_q::Ring consumer(cfg, false);
  std::vector<std::byte> payload(static_cast<std::size_t>(payload_bytes));
  std::vector<std::uint64_t> latencies;
  latencies.reserve(static_cast<std::size_t>(count));

  std::atomic<bool> go{false};
  std::atomic<int> done{0};

  std::printf("smh_q bench_throughput (count=%d payload=%dB)\n", count, payload_bytes);
  std::printf("mode         payload  count    elapsed_ms    throughput\n");
  std::printf("----------------------------------------------------------------\n");

  const auto t0 = Clock::now();

  std::thread reader([&] {
    while (!go.load(std::memory_order_acquire)) std::this_thread::yield();
    while (done.load(std::memory_order_relaxed) < count) {
      auto msg = consumer.try_consume();
      if (!msg.has_value()) {
        consumer.wait_readable(std::chrono::milliseconds(1));
        continue;
      }
      std::uint64_t pub = 0;
      std::memcpy(&pub, msg->data(), sizeof(pub));
      latencies.push_back(bench_now_ns() - pub);
      done.fetch_add(1, std::memory_order_relaxed);
    }
  });

  go.store(true, std::memory_order_release);

  for (int i = 0; i < count; ++i) {
    const std::uint64_t stamp = bench_now_ns();
    std::memcpy(payload.data(), &stamp, sizeof(stamp));
    while (!producer.try_publish(payload)) std::this_thread::yield();
  }

  reader.join();
  const auto t1 = Clock::now();
  smh_q::Ring::unlink(name);

  const double elapsed_ms =
      std::chrono::duration<double, std::milli>(t1 - t0).count();
  std::sort(latencies.begin(), latencies.end());

  const double msgs_per_sec = static_cast<double>(count) / (elapsed_ms / 1000.0);
  const double mb_per_sec =
      msgs_per_sec * static_cast<double>(payload_bytes) / (1024.0 * 1024.0);
  const double p50 =
      latencies.empty() ? 0.0 : static_cast<double>(latencies[latencies.size() / 2]);
  const double p99 = latencies.empty()
                         ? 0.0
                         : static_cast<double>(latencies[static_cast<std::size_t>(
                               0.99 * static_cast<double>(latencies.size() - 1))]);

  std::printf(
      "%-12s payload=%4dB count=%7d elapsed=%8.2f ms  msgs/s=%10.0f  MB/s=%8.2f  "
      "p50=%6.0f ns  p99=%6.0f ns\n",
      "threaded", payload_bytes, count, elapsed_ms, msgs_per_sec, mb_per_sec, p50, p99);

  return 0;
}

File: packages/picoipc/benchmarks/python/bench_throughput.py
#!/usr/bin/env python3
"""Throughput benchmark: threaded SPSC (paired threads)."""

import argparse
import os
import struct
import sys
import threading
import time

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from picoipc import Ring


def slot_bytes(payload_bytes: int) -> int:
    return max(64, payload_bytes + 8)


def percentile(samples: list[int], p: float) -> float:
    if not samples:
        return 0.0
    idx = int(p * (len(samples) - 1))
    return float(samples[idx])


def run_threaded(count: int, payload_bytes: int) -> None:
    name = f"smh_q_py_bench_tp_{os.getpid()}"
    Ring.unlink(name)

    producer = Ring(
        name,
        create=True,
        slot_count=256,
        slot_size=slot_bytes(payload_bytes),
    )
    consumer = Ring(name=name, create=False)

    payload = bytes(payload_bytes)
    latencies: list[int] = [0] * count
    received = 0
    go = threading.Event()

    def reader() -> None:
        nonlocal received
        go.wait()
        while received < count:
            msg = consumer.try_consume()
            if msg is None:
                continue
            pub_ns = struct.unpack_from("<Q", msg, 0)[0]
            latencies[received] = time.time_ns() - pub_ns
            received += 1

    t = threading.Thread(target=reader, daemon=True)
    t.start()

    t0 = time.perf_counter()
    go.set()
    for _ in range(count):
        stamp = struct.pack("<Q", time.time_ns())
        body = stamp + payload[len(stamp):]
        while not producer.try_publish(body):
            pass

    t.join(timeout=30.0)
    t1 = time.perf_counter()
    producer.close(unlink=True)

    elapsed_ms = (t1 - t0) * 1000.0
    samples = sorted(latencies[:received])
    msgs_per_sec = count / (elapsed_ms / 1000.0)
    mb_per_sec = msgs_per_sec * payload_bytes / (1024.0 * 1024.0)
    print(
        f"{'threaded':<12} payload={payload_bytes:4d}B count={count:7d} "
        f"elapsed={elapsed_ms:8.2f} ms  msgs/s={msgs_per_sec:10.0f}  "
        f"MB/s={mb_per_sec:8.2f}  p50={percentile(samples, 0.50):6.0f} ns  "
        f"p99={percentile(samples, 0.99):6.0f} ns"
    )


def run_sequential(count: int, payload_bytes: int) -> None:
    name = f"smh_q_py_bench_seq_{os.getpid()}"
    Ring.unlink(name)

    ring = Ring(
        name,
        create=True,
        slot_count=max(count, 256),
        slot_size=slot_bytes(payload_bytes),
    )
    payload = bytes(payload_bytes)

    t0 = time.perf_counter()
    for _ in range(count):
        while not ring.try_publish(payload):
            pass
    for _ in range(count):
        while ring.try_consume() is None:
            pass
    t1 = time.perf_counter()
    ring.close(unlink=True)

    elapsed_ms = (t1 - t0) * 1000.0
    msgs_per_sec = count / (elapsed_ms / 1000.0)
    mb_per_sec = msgs_per_sec * payload_bytes / (1024.0 * 1024.0)
    print(
        f"{'sequential':<12} payload={payload_bytes:4d}B count={count:7d} "
        f"elapsed={elapsed_ms:8.2f} ms  msgs/s={msgs_per_sec:10.0f}  "
        f"MB/s={mb_per_sec:8.2f}"
    )


def main() -> int:
    p = argparse.ArgumentParser(description="smh_q Python throughput benchmark")
    p.add_argument("count", nargs="?", type=int, default=100_000)
    p.add_argument("payload", nargs="?", type=int, default=0,
                   help="payload bytes (0 = run 64, 256, 1024)")
    args = p.parse_args()

    payloads = [args.payload] if args.payload > 0 else [64, 256, 1024]

    print(f"smh_q bench_throughput (count={args.count})")
    print("mode         payload  count    elapsed_ms    throughput")
    print("----------------------------------------------------------------")

    for pb in payloads:
        run_threaded(args.count, pb)
        run_sequential(args.count, pb)
        print()

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
